### DISCUSSION TITLE
Lesson 11: NFT AttributeError: Final argument must be a dict of transaction parameters that includes a `from` field specifying the address to deploy from

### DISCUSSION THREAD

**Discussion by sevamove** [smartcontractkit/full-blockchain-solidity-course-py#383] [Q&A] [answered]
I came across the following error when running cmd `brownie run scripts/deploy_and_create.py --network rinkeby` on [10:18:23](https://youtu.be/M576WGiDBdQ?t=37101)  time line of the tutorial:
```
mbp:nft-demo mac$ brownie run scripts/deploy_and_create.py --network rinkeby
Brownie v1.16.4 - Python development framework for Ethereum

NftDemoProject is the active project.

Running 'scripts/deploy_and_create.py::main'...
  File "brownie/_cli/run.py", line 49, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/deploy_and_create.py", line 12, in main
    simple_collectible = SimpleCollectible.deploy({"from": account})
  File "brownie/network/contract.py", line 595, in __call__
    raise AttributeError(
AttributeError: Final argument must be a dict of transaction parameters that includes a `from` field specifying the address to deploy from
mbp:nft-demo mac$ 

```

Found 0 additional information about the error on the internet. 
****

Here is `brownie-config.yaml`:
```yaml
dependencies:
    # taken from https://docs.openzeppelin.com/contracts/3.x/erc721#constructing_an_erc721_token_contract
  - OpenZeppelin/openzeppelin-contracts@3.4.0
compiler:
  solc:
    remappings:
      - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@3.4.0'
dotenv: .env
wallets:
  from_key: ${PRIVATE_KEY}
```
Here is `contracts/SimpleCollectible.sol`:
```solidity
// SPDX-Licence-Identifier: MIT

pragma solidity 0.6.6; // also works fine with the ^0.8.0 version.

import "@openzeppelin/contracts/token/ERC721/ERC721.sol";

contract SimpleCollectible is ERC721 {
    uint256 public tokenCounter;
    
    constructor () public ERC721 ("Dogie", "DOG") {
        tokenCounter = 0;
    }


    function createCollectible(string memory tokenURI) public returns (uint256) {
        uint256 newTokenId = tokenCounter;
        _safeMint(msg.sender, newTokenId);
        _setTokenURI(newTokenId, tokenURI);
        tokenCounter = tokenCounter + 1;
        return newTokenId;
    }
}

```
Here is `scripts/deploy_and_create.py` script:
```py
from scripts.helpful_scripts import get_account
from brownie import SimpleCollectible

# If you can not see through this link you need to add an extension to your browser like IPFS Companion.
# tokenURI
sample_token_uri = "https://ipfs.io/ipfs/Qmd9MCGtdVz2miNumBHDbvj8bigSgTwnr4SbyH6DNnpWdt?filename=0-PUG.json"
# In order to view image of NFT.
OPENSEA_URL = "https://testnets.opensea.io/assets/{}/{}"

def main():
    account = get_account()
    simple_collectible = SimpleCollectible.deploy({"from": account})
    tx = simple_collectible.createCollectible(sample_token_uri, {"from": account})
    tx.wait(1)
    print(
        f"Awesome, you can view your NFT at {OPENSEA_URL.format(simple_collectible.address, simple_collectible.tokenCounter() - 1)}"
    )
    print("Please wait up to 20 minutes, and hit the refresh metadata button.")

```
`scripts/helpful_scripts.py`:
```py
from brownie import accounts, network, config

LOCAL_BLOCKCHAIN_ENVIRONMENTS = [
    "development",
    "ganache",
    "hardhat",
    "local-ganache",
    "mainnet-fork",
]

def get_account(index=None, id=None):
    if index:
        return accounts[index]
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        return accounts[0]
    if id:
        return accounts.load(id)
    if network.show_active() in config["networks"]:
        return accounts.add(config["wallets"]["from_key"])
    return None
```
And an example of `.env`:
```bash
export PRIVATE_KEY=0xaxaxa
export WEB3_INFURA_PROJECT_ID=a7ceaxaxax
export ETHERSCAN_TOKEN=xxxaxax

```
Glad to hear any suggestions 😁

**Comment by cromewar** [comment, marked as answer]
Hello @moonwake769 this is extremely weird because your syntax appears to be ok as you can see [here](https://eth-brownie.readthedocs.io/en/stable/core-contracts.html#transaction-parameters) so I think something is wrong with your brownie.. Please try this:

1. create a new virtualenv for python and install eth-brownie on it.
2. activate the virutalenv
3. delete the build folder and compile again.
4. try to run the code again.

**Reply by sevamove** [reply]
Hello @cromewar, thanks for your reaction. I tried to follow the steps as you recommended by running the commands in `nft-demo` folder:
```sh
$ python3 -m venv venv
$ source venv/bin/activate
$ pip3 install eth-brownie
$ rm -r build
$ brownie compile
$ brownie run scripts/deploy_and_create.py --network rinkeby
```
****
Unfortunately, the same `AttributeError` still occurs.

**Reply by cromewar** [reply]
I got the same error when the `account` does not exists as you can see the example bellow

![image](https://user-images.githubusercontent.com/20199180/142225373-2ea34650-15f8-4798-bb07-22856e9baa6b.png)

I seems your brownie is ignoring completely the statement `{"from": account}` this makes me wonder if your get_account is actually working properly, could you copy paste your helpful scripts here please?

**Reply by sevamove** [reply]
Of course, just updated my first message in this discussion 👍
For convenience, here is the copy of it:
```py
from brownie import accounts, network, config

LOCAL_BLOCKCHAIN_ENVIRONMENTS = [
    "development",
    "ganache",
    "hardhat",
    "local-ganache",
    "mainnet-fork",
]

def get_account(index=None, id=None):
    if index:
        return accounts[index]
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        return accounts[0]
    if id:
        return accounts.load(id)
    if network.show_active() in config["networks"]:
        return accounts.add(config["wallets"]["from_key"])
    return None
```

**Reply by sevamove** [reply]
In addition to it, I just tried to run `brownie run scripts/deploy_and_create.py` by deploying the contract on `ganache` network. Deployed successfully. So the problem might not be with the `brownie`.

**Reply by cromewar** [reply]
Ok now I know what the problem exactly is, look you are trying to develop to Rinkeby, which is not in `LOCAL_BLOCKCHAIN_ENVIRONMENTS`, you are not providing neither and id or an index, so the return is:

```Python
if network.show_active() in config["networks"]:
        return accounts.add(config["wallets"]["from_key"])
```

But the if statement says if the network is on your brownie-config, and as you can see you don't have rinkeby on your brownie-config:

![image](https://user-images.githubusercontent.com/20199180/142236724-8a78558b-4ef6-46d5-bc83-925804989cbc.png)

so the `get_account` function is always returning `None` and that's why you have this error.

You must add rinkeby to the networks, something like this: 

```YAML
rinkeby:
    verify: True
```
Or you can change the return of get_account as well.

give it a try.

**Reply by sevamove** [reply]
Nice catch 😸 You are right about `helpful_scripts.py`. I watched the video not carefully and missed the last `return` statement by simply copy-past the function from `aave-brownie` lesson. So I changed the `get_account()` function:

from:
```py
def get_account(index=None, id=None):
    if index:
        return accounts[index]
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        return accounts[0]
    if id:
        return accounts.load(id)
    if network.show_active() in config["networks"]:
        return accounts.add(config["wallets"]["from_key"])
    return None
```
to:
```py
def get_account(index=None, id=None):
    if index:
        return accounts[index]
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        return accounts[0]
    if id:
        return accounts.load(id)
    return accounts.add(config["wallets"]["from_key"])
```
Thanks a lot @cromewar 

> Lesson from this discussion: always watch carefully every second from any tutorial.

**Reply by cromewar** [reply]
I'm glad it helped ;)
